Makefile

```diff
@@ -1,4 +1,17 @@
 # ---------------------------------------------------------------------
+# Build configuration
+# ---------------------------------------------------------------------
+
+BUILD_DIRECTORY ?= release
+
+ifndef BUILD_DIRECTORY
```

Comment:
Hmmm, that just changes `make clean` to delete everything **in** the BUILD_DIRECTORY, rather than deleting the BUILD_DIRECTORY itself. I'm not sure that's much better :-/

IMHO it _should_ be valid to do `BUILD_DIRECTORY=/tmp make my-target`, but `BUILD_DIRECTORY=/tmp make clean` _shouldn't_ do `rm -rf /tmp/*` (it deletes more than it should), instead it should only delete files that we know the `Makefile` (and associated scripts) created. I'll have a think tomorrow (when I'm less tired) and see if I can come up with a nice solution.


Of course the alternative would be to do:
```Makefile
BUILD_DIRECTORY_PARENT ?= .
BUILD_DIRECTORY = $(BUILD_DIRECTORY_PARENT)/release
```
and then allow the user to (optionally) do `BUILD_DIRECTORY_PARENT=/tmp make my-target`, which means that everything would get built inside `/tmp/release`, and the 'clean' rule could stay as `rm -rf $(BUILD_DIRECTORY)`, but I'm not sure if that feels very nice...